### DISCUSSION TITLE
Best practice about migration？

### DISCUSSION THREAD

**Discussion by suerta-git** [mikro-orm/mikro-orm#2617] [Q&A] [answered]
Hi folks, I am new for mikro and learning to use it in my own application.

Current state:
- I have prepared my entities
- Have no migration files yet
- Production DB is empty now

Expected:
- Apply migrations to production DB
- Can migrate automaticlly by latest migration files, rather than execute some command or script manually

My idea:
- Prepare migrarion files using CLI, and deploy them together with codes to remote
- Migrate programmatically after every time app started and orm initialized

What's your idea about it? 

Because my application is built on **serverless** so I concern that maybe it will cold start frequently so that `migrator.up()` frequently. Don't know if it will cause performance issues.

**Comment by B4nan** [comment, marked as answer]
Seems ok to me, but I have no experience with serverless apps, so there might be better way to do this. Maybe you could at least implement some state cache, so you would only run migrations if you detect new app version (e.g. latest commit hash is different from the one stored in your state cache). As there is no reason to run migrations over and over, only if you deploy new version of your app.

One way to speed this up could be using `migrationsList`: 
https://mikro-orm.io/docs/migrations#importing-migrations-statically

As otherwise migrator will check the file system and usually file system operations are taking a lot. But you will need to measure yourself if it has a real impact.

**Reply by suerta-git** [reply]
Thanks for your opinion. I expect there is an out-of-box cache to avoid unnecessary migration before but seems no.
And now I have wrote a migrating function and only trigger it if there is a new commit on migration files. I think it's good enough for current stage.

**Reply by B4nan** [reply]
> I expect there is an out-of-box cache to avoid unnecessary migration before but seems no.

We do check if migration was executed and execute only the pending ones, that obviously work. I am talking about how you can optimize things, as it is quite a lot of work that should definitely not happen before each request is handled.

**Reply by suerta-git** [reply]
Besides, it's awesome if I can import migration configurations from an api rather than file system. Not sure is it possible.

**Reply by B4nan** [reply]
I already shared the link about `migrationsList`, I believe you could feed the migrations dynamically from any source into it.

**Reply by suerta-git** [reply]
Sorry I'm still a little unclear about loading migrations dynamically 😂.
I have read your link. Does that mean I should compile and import classes from downloaded file contents (maybe plain texts) then give them to `migrationsList`? 🤔
